app/controllers/concerns/authenticates_with_two_factor.rb

```diff
@@ -53,16 +56,50 @@ def authenticate_with_two_factor
 
   private def authenticate_with_two_factor_via_otp(user)
     if valid_otp_attempt?(user) || valid_backup_code_attempt?(user)
-      # Remove any lingering user data from login
-      session.delete(:otp_user_id)
+      # add 2fa device if true
+      browser = Browser.new(request.user_agent)
+      # foce a device name, even if none provided
+      device_name = user_params[:device_name].presence || "#{browser.name} #{browser.platform.name}"
 
-      user.save!
-      sign_in(user, message: :two_factor_authenticated, event: :authentication)
+      add_2fa_device(user, device_name) if user_params[:remember_device] == 'true' && bypass_2fa_enabled?
```

Comment:
what happens if we drop the connection between add_2fa_devise and two_factor_successful. If the  use retries i think we end up with duplicate remembered devices. Do we care? can we do anything about it?

Reply:
In the case where two_factor_successful fails after add_2fa_device, a device would still be added meaning the user would be able to log in again without needing to input 2fa on the next time. I'll move add_2fa_device to be after sign_in for now, meaning a device would not be added until the user has successfully signed in.